[Twitter] New function allowing to tweet with an attached picture.

SR-G opened the pull request:
Old methods are working like previously.

New method allows to send a tweet with an attached picture - picture either taken from a remote URL (file content is then grabbed locally and deleted once tweet is sent), either taken from a local file.

Result is like in : https://i.imgur.com/F9eDwMY.png
Topic : https://community.openhab.org/t/twitter-action-how-to-send-a-tweet-containing-an-image/35623

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -8,14 +8,24 @@
+import org.apache.commons.io.FilenameUtils;
+
 import org.apache.commons.lang.StringUtils;
+import org.apache.commons.io.FileUtils;
```
Duplicate of line 16 above.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+    }
+    
+    /**
+     * Internal method about sending a tweet, with or without image
```
"for sending", not "about sending"

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
             return true;
         } catch (TwitterException e) {
-            logger.error("Failed to send Tweet '" + tweetTxt + "' because of: " + e.getLocalizedMessage());
+            logger.error("Failed to send Tweet '" + abbreviatedTweetTxt + "' because of: " + e.getLocalizedMessage());
```
This should be a WARN instead of an ERROR.
Also, need to use parameterized logging instead of string concatenation.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -40,15 +50,12 @@
      */
-    @ActionDoc(text = "Sends a Tweet via Twitter", returns = "<code>true</code>, if sending the tweet has been successful and <code>false</code> in all other cases.")
-    public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet to send") String tweetTxt) {
+    private static boolean checkPrerequesites() {
```
"checkPrerequisites"

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
     }
 
     /**
+     * Sends a Tweet via Twitter
```
I suggest this should say "Sends a standard Tweet."

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+    }
+
+    /**
+     * Sends a Tweet via Twitter
```
I suggest this should say "Sends a Tweet with an image."

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+    /**
+     * Sends a Tweet via Twitter
+     * 
+     * @param tweetTxt the Tweet to send
```
This method has two parameters, but only one is listed in the Javadoc.
Also, the "tweetTxt" parameter here is not a "Tweet to send", but a text string to be sent as a Tweet.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+        // prepare the image attachment
+        File fileToAttach = null;
+        boolean deleteTemporaryFile = false;
+        if (StringUtils.startsWith(tweetPicture, "http")) {
```
Need to skip this check. What if I want to send an attachment named "http.jpg" ?

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+        		// fileToAttach.deleteOnExit();  // good idea ? could lead to temporary files staying around until JVM is down
+        		deleteTemporaryFile = true;
+        		FileUtils.copyURLToFile(url, fileToAttach);
+        	} catch (final MalformedURLException e) {
```
Not final.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+        		deleteTemporaryFile = true;
+        		FileUtils.copyURLToFile(url, fileToAttach);
+        	} catch (final MalformedURLException e) {
+        		logger.error("Can't read file from '" + tweetPicture + "'", e);
```
This should be a WARN instead of an ERROR.
Also, need to use parameterized logging instead of string concatenation.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+        		FileUtils.copyURLToFile(url, fileToAttach);
+        	} catch (final MalformedURLException e) {
+        		logger.error("Can't read file from '" + tweetPicture + "'", e);
+        	} catch (final IOException e) {
```
Not final.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+        	} catch (final MalformedURLException e) {
+        		logger.error("Can't read file from '" + tweetPicture + "'", e);
+        	} catch (final IOException e) {
+        		logger.error("Can't save file from '" + tweetPicture + "' to '" + path + "'", e);
```
This should be a WARN instead of an ERROR.
Also, need to use parameterized logging instead of string concatenation.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+        }
+    
+        // send the Tweet
+        final boolean result = sendTweet(tweetTxt, fileToAttach);
```
Not final.

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -40,15 +50,12 @@
 
     /**
-     * Sends a Tweet via Twitter
+     * Check twitter prerequesites. Should be used at the beginning of all public methods
```
"prerequisites"

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -40,15 +50,12 @@
-     * 
-     * @return <code>true</code>, if sending the tweet has been successful and
-     *         <code>false</code> in all other cases.
+     * @return <code>true</code>, all prerequesites are validated and 
```
"prerequisites"

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -40,15 +50,12 @@
-     * @return <code>true</code>, if sending the tweet has been successful and
-     *         <code>false</code> in all other cases.
+     * @return <code>true</code>, all prerequesites are validated and 
+     *         <code>false</code> one prerequesite is not validated.
```
"prerequisite"

9037568 commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -57,21 +64,104 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+     *         <code>false</code> in all other cases.
+     */
+    @ActionDoc(text = "Sends a Tweet via Twitter", returns = "<code>true</code>, if sending the tweet has been successful and <code>false</code> in all other cases.")
+    public static boolean sendPicture(@ParamDoc(name = "tweetTxt", text = "the Tweet to send") String tweetTxt, @ParamDoc(name = "tweetPicture", text = "the picture to attach") String tweetPicture) {
```
I suggest this method should be named "sendTweet".

SR-G commented:
Asked corrections have been applyed ("prerequisite" typo, logger.warn, ...).
Test about parameter being either an image has been updated in order to have a better detection, but the test needs to be kept (as the behavior is not the same if the parameter is a remote URL or a local file).

SR-G commented on bundles/action/org.openhab.action.twitter/src/main/java/org/openhab/action/twitter/internal/Twitter.java:
```diff
@@ -58,34 +64,130 @@ public static boolean sendTweet(@ParamDoc(name = "tweetTxt", text = "the Tweet t
+     * @return <code>true</code>, if sending the tweet has been successful and
+     *         <code>false</code> in all other cases.
+     */
+    private static boolean doSendTweet(final String tweetTxt, final File fileToAttach) {
```
Can't be renamed that way as it would generate a signature collision. Moreover is a private internal method, not to be used in rules.

9037568 replied:
Of course it can. You just have to cast your nulls appropriately:

Line 114:

````
return sendTweet(tweetTxt, (File)null);
````

Another option would be to make the File parameter optional so that you can avoid sending null altogether.

kaikreuzer commented:
@9037568 Why did you add the "stale" label on this? It looks as if the last activity were new commits that addressed your last comments.

9037568 commented:
The request for changes to the method signature on line 81 wasn't addressed.

SR-G commented:
@9037568 you do understand that the way you are handling that kind of push (with a lot of very, very, very minor remarks and for some of them quite arguable) is a real break for new joiners in the project to be willing to participate any further ?
The commited code was working since the first commit. I've addressed most of your remarks. Many many months after that we are still in the same place. That's enough for me. So bad !

9037568 commented:
I'm terribly sorry you find the process difficult.
There, I fixed it for you.
